satpy/scene.py

```diff
@@ -1247,147 +954,447 @@ def to_xarray_dataset(self, datasets=None):
+                            patterns that will be filled in by dataset
+                            attributes.
+            datasets (iterable): Limit written products to these datasets
+            compute (bool): If `True` (default), compute all of the saves to
+                disk. If `False` then the return value is either a
+                :doc:`dask:delayed` object or two lists to be passed to
+                a `dask.array.store` call. See return values below for more
+                details.
+            kwargs: Additional writer arguments. See :doc:`../writers` for more
+                information.
+
+        Returns:
+            Value returned depends on `compute` keyword argument. If
+            `compute` is `True` the value is the result of a either a
+            `dask.array.store` operation or a :doc:`dask:delayed`
+            compute, typically this is `None`. If `compute` is `False` then the
+            result is either a :doc:`dask:delayed` object that can be
+            computed with `delayed.compute()` or a two element tuple of
+            sources and targets to be passed to :func:`dask.array.store`. If
+            `targets` is provided then it is the caller's responsibility to
+            close any objects that have a "close" method.
+
+        """
+        dataarrays = self._get_actual_dataarrays(datasets)
+        if not dataarrays:
+            raise RuntimeError("None of the requested datasets have been "
+                               "generated or could not be loaded. Requested "
+                               "composite inputs may need to have matching "
+                               "dimensions (eg. through resampling).")
+        if writer is None:
+            if filename is None:
+                writer = 'geotiff'
+            else:
+                writer = self._get_writer_by_ext(os.path.splitext(filename)[1])
+        writer, save_kwargs = load_writer(writer,
+                                          ppp_config_dir=self._ppp_config_dir,
+                                          filename=filename,
+                                          **kwargs)
+        return writer.save_datasets(dataarrays, compute=compute, **save_kwargs)
+
+    @staticmethod
+    def _get_writer_by_ext(extension):
+        """Find the writer matching the ``extension``.
+
+        Defaults to "simple_image".
+
+        Example Mapping:
+
+            - geotiff: .tif, .tiff
+            - cf: .nc
+            - mitiff: .mitiff
+            - simple_image: .png, .jpeg, .jpg, ...
+
+        Args:
+            extension (str): Filename extension starting with
+                "." (ex. ".png").
+
+        Returns:
+            str: The name of the writer to use for this extension.
+
+        """
+        mapping = {".tiff": "geotiff", ".tif": "geotiff", ".nc": "cf",
+                   ".mitiff": "mitiff"}
+        return mapping.get(extension.lower(), 'simple_image')
+
+    def remove_failed_datasets(self, keepables):
```

Comment:
Did this have to be made public for the case where we call it above on `new_scn`?

Reply:
yes, exactly. As you know, it's bad practice to use the private methods of a class outside of it.